File: src/web/src/constants/api.constants.ts
/**
 * @fileoverview API Constants for Smart Apparel System
 * @version 1.0.0
 * 
 * Contains core API-related constants including endpoints, headers,
 * request configurations, and WebSocket settings for the smart-apparel
 * web application.
 */

/**
 * Current API version identifier
 */
export const API_VERSION = 'v1';

/**
 * API endpoint configurations organized by domain
 */
export const API_ENDPOINTS = {
  AUTH: {
    LOGIN: '/auth/login',
    SIGNUP: '/auth/signup',
    REFRESH: '/auth/refresh',
    LOGOUT: '/auth/logout',
    VERIFY: '/auth/verify',
    RESET_PASSWORD: '/auth/reset-password',
    MFA: '/auth/mfa'
  },
  SENSOR: {
    CALIBRATE: '/sensor/calibrate',
    DATA: '/sensor/data',
    STATUS: '/sensor/status',
    CONFIGURE: '/sensor/configure',
    DIAGNOSTICS: '/sensor/diagnostics',
    FIRMWARE: '/sensor/firmware'
  },
  SESSION: {
    CREATE: '/session/create',
    UPDATE: '/session/update',
    END: '/session/end',
    LIST: '/session/list',
    DETAILS: '/session/details',
    EXPORT: '/session/export',
    SHARE: '/session/share'
  },
  ATHLETE: {
    PROFILE: '/athlete/profile',
    METRICS: '/athlete/metrics',
    HISTORY: '/athlete/history',
    GOALS: '/athlete/goals',
    PROGRESS: '/athlete/progress',
    RECOMMENDATIONS: '/athlete/recommendations'
  },
  TEAM: {
    LIST: '/team/list',
    MEMBERS: '/team/members',
    STATS: '/team/stats',
    SCHEDULE: '/team/schedule',
    PERFORMANCE: '/team/performance',
    COMPARISON: '/team/comparison'
  },
  ALERT: {
    LIST: '/alert/list',
    CREATE: '/alert/create',
    UPDATE: '/alert/update',
    DELETE: '/alert/delete',
    SETTINGS: '/alert/settings',
    HISTORY: '/alert/history'
  },
  ANALYTICS: {
    HEATMAP: '/analytics/heatmap',
    PERFORMANCE: '/analytics/performance',
    BIOMECHANICS: '/analytics/biomechanics',
    TRENDS: '/analytics/trends',
    REPORTS: '/analytics/reports',
    EXPORT: '/analytics/export'
  }
} as const;

/**
 * Standard HTTP headers for API requests
 */
export const API_HEADERS = {
  CONTENT_TYPE: 'application/json',
  AUTHORIZATION: 'Authorization',
  ACCEPT: 'application/json',
  API_KEY: 'X-API-Key',
  CORRELATION_ID: 'X-Correlation-ID',
  CLIENT_VERSION: 'X-Client-Version'
} as const;

/**
 * API request configuration parameters
 */
export const REQUEST_CONFIG = {
  /** Request timeout in milliseconds */
  TIMEOUT: 30000,
  /** Number of retry attempts for failed requests */
  RETRY_ATTEMPTS: 3,
  /** Delay between retry attempts in milliseconds */
  RETRY_DELAY: 1000,
  /** Rate limiting configurations per endpoint type */
  RATE_LIMIT: {
    REAL_TIME: 100, // requests per second
    QUERY: 1000, // requests per minute
    INTEGRATION: 500, // requests per minute
    ANALYTICS: 200, // requests per minute
    EXPORT: 50 // requests per minute
  },
  /** Maximum number of items in batch requests */
  BATCH_SIZE: 100,
  /** Cache duration in milliseconds */
  CACHE_DURATION: 300000
} as const;

/**
 * WebSocket configuration for real-time data streaming
 */
export const WEBSOCKET_CONFIG = {
  /** Maximum number of reconnection attempts */
  RECONNECT_ATTEMPTS: 5,
  /** Interval between reconnection attempts in milliseconds */
  RECONNECT_INTERVAL: 5000,
  /** Interval for ping messages in milliseconds */
  PING_INTERVAL: 30000,
  /** Timeout for heartbeat messages in milliseconds */
  HEARTBEAT_TIMEOUT: 60000,
  /** Maximum message size in bytes (1MB) */
  MESSAGE_SIZE_LIMIT: 1048576,
  /** Threshold for message compression in bytes (1KB) */
  COMPRESSION_THRESHOLD: 1024
} as const;

// Type definitions for better TypeScript support
export type ApiEndpoints = typeof API_ENDPOINTS;
export type ApiHeaders = typeof API_HEADERS;
export type RequestConfig = typeof REQUEST_CONFIG;
export type WebSocketConfig = typeof WEBSOCKET_CONFIG;

File: src/ios/SmartApparel/Sources/Constants/APIConstants.swift
import Foundation // v13.0+

/// Provides centralized access to all API-related configuration constants throughout the iOS application.
/// This struct contains comprehensive definitions for API endpoints, headers, timeouts, and other
/// networking-related configuration parameters supporting both REST and GraphQL APIs.
public struct APIConstants {
    
    // MARK: - Base Configuration
    
    /// Base URL for all API endpoints
    public static let BASE_URL = "https://api.smartapparel.com"
    
    /// Current API version
    public static let API_VERSION = "v1"
    
    /// WebSocket URL for real-time data streaming
    public static let WEBSOCKET_URL = "wss://api.smartapparel.com/ws"
    
    /// GraphQL endpoint for complex queries
    public static let GRAPHQL_ENDPOINT = "/graphql"
    
    // MARK: - API Endpoints
    
    /// Dictionary containing all API endpoint paths organized by domain
    public static let API_ENDPOINTS: [String: [String: String]] = [
        "AUTH": [
            "LOGIN": "/auth/login",
            "SIGNUP": "/auth/signup",
            "REFRESH": "/auth/refresh",
            "LOGOUT": "/auth/logout",
            "MFA": "/auth/mfa",
            "BIOMETRIC": "/auth/biometric"
        ],
        "SENSOR": [
            "DATA": "/sensor/data",
            "CALIBRATION": "/sensor/calibrate",
            "STATUS": "/sensor/status",
            "STREAM": "/sensor/stream",
            "BATCH": "/sensor/batch",
            "HEALTH": "/sensor/health"
        ],
        "SESSION": [
            "CREATE": "/session/create",
            "UPDATE": "/session/update",
            "END": "/session/end",
            "LIST": "/session/list",
            "EXPORT": "/session/export",
            "SHARE": "/session/share"
        ],
        "ANALYTICS": [
            "PERFORMANCE": "/analytics/performance",
            "HEATMAP": "/analytics/heatmap",
            "METRICS": "/analytics/metrics",
            "EXPORT": "/analytics/export",
            "TRENDS": "/analytics/trends",
            "PREDICTIONS": "/analytics/predictions"
        ],
        "TEAM": [
            "LIST": "/team/list",
            "DETAILS": "/team/details",
            "MEMBERS": "/team/members",
            "STATS": "/team/stats",
            "PERMISSIONS": "/team/permissions",
            "INVITES": "/team/invites"
        ]
    ]
    
    // MARK: - HTTP Headers
    
    /// Standard HTTP headers used in API requests
    public static let HTTP_HEADERS: [String: String] = [
        "CONTENT_TYPE": "application/json",
        "ACCEPT": "application/json",
        "API_KEY": "X-API-Key",
        "AUTH_TOKEN": "Authorization",
        "DEVICE_ID": "X-Device-ID",
        "APP_VERSION": "X-App-Version",
        "CORRELATION_ID": "X-Correlation-ID",
        "REQUEST_ID": "X-Request-ID",
        "SIGNATURE": "X-Signature",
        "TIMESTAMP": "X-Timestamp"
    ]
    
    // MARK: - Timeout Configuration
    
    /// Timeout intervals for different types of network operations (in seconds)
    public static let TIMEOUT_INTERVALS: [String: TimeInterval] = [
        "REQUEST": 30.0,      // Standard API request timeout
        "RESOURCE": 300.0,    // Resource loading timeout
        "WEBSOCKET": 30.0,    // WebSocket connection timeout
        "UPLOAD": 600.0,      // File upload timeout
        "DOWNLOAD": 600.0,    // File download timeout
        "STREAM": 3600.0      // Long-running stream timeout
    ]
    
    // MARK: - Error Codes
    
    /// Standard HTTP error codes used in the application
    public static let ERROR_CODES: [String: Int] = [
        "UNAUTHORIZED": 401,
        "FORBIDDEN": 403,
        "NOT_FOUND": 404,
        "RATE_LIMIT": 429,
        "SERVER_ERROR": 500,
        "SERVICE_UNAVAILABLE": 503,
        "GATEWAY_TIMEOUT": 504
    ]
    
    // MARK: - Retry Configuration
    
    /// Configuration parameters for request retry behavior
    public static let RETRY_CONFIG: [String: Double] = [
        "MAX_ATTEMPTS": 3,            // Maximum number of retry attempts
        "BACKOFF_MULTIPLIER": 2.0,    // Exponential backoff multiplier
        "INITIAL_DELAY": 1.0,         // Initial retry delay in seconds
        "MAX_DELAY": 60.0,           // Maximum retry delay in seconds
        "JITTER": 0.1                // Random jitter factor for retry timing
    ]
}